### PR TITLE
MRG: add plugin support for new command-line subcommands

### PR DESCRIPTION
**Pull request by ctb**
Add command-line plug-in interface, per https://github.com/sourmash-bio/sourmash/issues/1353.

This PR builds on https://github.com/sourmash-bio/sourmash/pull/2428 to provide support for adding new sourmash subcommands, under `sourmash scripts <cmd>`.

The CLI plugin interface looks like this on the client side -
```
class cmd_somewhat:
    command = 'somewhat'
    description = "does a thing"

    def __init__(self, subparser):
        super().__init__(p)
        # add arguments etc here 
        debug_literal('RUNNING cmd_somewhat.__init__')
        subparser.add_argument('foo', type=int)

    def main(self, args):
        super().main(args)
        # what we actually run. 
        print('RUNNING cmd', self, args)
        print('XYZ somewhat', args.foo)
        print(f'2*{args.foo} is {2*args.foo}')
```

Adding this plugin enables:
```
% sourmash scripts  -h          

== This is sourmash version 4.6.1. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

usage:  scripts [-h]

optional arguments:
  -h, --help  show this help message and exit

extension commands:
  sourmash scripts somewhat --help - do a thing
```
and
```
% sourmash scripts somewhat 5 

== This is sourmash version 4.6.1. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

...
2*5 is 10
```
which is prety cool 🎉 

## Alternatives to `script`

could be `sourmash ext` or `sourmash plugin`?

### CODE REVIEW

**Comment by ctb** [comment]
Notes to self/questions for review: is `scripts` the right subcommand to use here?

**Comment by ctb** [comment]
Ready for review!

File: doc/dev_plugins.md [hunk 1]
```diff
@@ -24,23 +25,35 @@ a_reader = "module_name:load_sketches"
...
 for saving and loading code already used in sourmash.
+* `Command_NewCommand` should be a class that subclasses
+  `plugins.CommandLinePlugin` and provides an `__init__` and
+  `main` method.
 
-Note that if the function or class has a `priority` attribute, this will
-be used to determine the order in which the plugins are called.
+Note that if the reader function or writer class has a `priority`
+attribute, this will be used to determine the order in which the
+plugins are called.
```
Comment (bluegenes) [inline doc/dev_plugins.md]: what guidelines do we want to have for priority? e.g. internal sourmash commands use priority xx - yy, external should use priority values above this, lower numbers are higher priority...

File: doc/dev_plugins.md [hunk 2]
```diff
@@ -53,15 +66,19 @@ Some (early stage) plugins are also available as examples:
 
 ## Debugging plugins
 
+`sourmash info -v` will list all detected plugins of any known kind.
```
Comment (bluegenes) [inline doc/dev_plugins.md]: "of any known kind" is a bit confusing to me.

maybe: "any detected plugins that are locally available" ?

File: src/sourmash/cli/scripts/__init__.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+src/sourmash/plugins.py for core sourmash implementation code, and
+https://github.com/sourmash-bio/sourmash_plugin_template for a template repo
+for making new plugins.
+"""
+
+# CTB TODO:
+# * provide suggestions for documentation & metadata for authors:
+# * provide guidance on how to test your CLI plugin at the CLI
+#   (minimal testing regime: sourmash scripts, look for description etc.)
+
```
Comment (bluegenes) [inline src/sourmash/cli/scripts/__init__.py:13]: later todo?

**Comment by bluegenes** [comment]
> Notes to self/questions for review: is scripts the right subcommand to use here?

I think you mentioned `sourmash ext` as an option somewhere. I really like that -- it's short and gets across the idea that the plugin is an "extension" beyond standard sourmash. But scripts is also good - maybe alias?

File: src/sourmash/cli/scripts/__init__.py (hunk 3 above)
Comment (ctb) [reply src/sourmash/cli/scripts/__init__.py:13]: yep, exactly.

File: doc/dev_plugins.md (hunk 2 above)
Comment (ctb) [reply doc/dev_plugins.md]: changed it to:
>all installed plugins

thanks ;)

Comment (ctb) [reply doc/dev_plugins.md]: in 7fa809ee

File: doc/dev_plugins.md (hunk 1 above)
Comment (ctb) [reply doc/dev_plugins.md]: adjusted in 6b034ae1. Still not perfect, but I think it needs more thought (and more examples) before we settle on numbers ;).

**Comment by ctb** [comment]
> > Notes to self/questions for review: is scripts the right subcommand to use here?
> 
> I think you mentioned `sourmash ext` as an option somewhere. I really like that -- it's short and gets across the idea that the plugin is an "extension" beyond standard sourmash. But scripts is also good - maybe alias?

added as an alias in 0813a0183da7e.

I think we should probably switch to it as the default, but I'm going to punt that to the next PR that will do more documenting and testing!

**Review by bluegenes — APPROVED** [review]
Lgtm!

**Pull request merged**